### PR TITLE
Switch internal scale type to autoscaling, enable apps/v1 scale subresources

### PR DESCRIPTION
**Pull request by liggitt**
xref #49504

* Switch workload internal scale type to autoscaling.Scale (internal-only change)
* Enable scale subresources for apps/v1 deployments, replicasets, statefulsets

```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/apps/v1beta1/conversion.go [hunk 1]
```diff
@@ -23,9 +23,11 @@ import (
 	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/kubernetes/pkg/apis/apps"
+	autoscaling "k8s.io/kubernetes/pkg/apis/autoscaling"
```
Comment (sttts) [inline pkg/apis/apps/v1beta1/conversion.go]: nit: no need for the alias

**Comment by liggitt** [comment]
cc @crimsonfaith91 @DirectXMan12

**Comment by crimsonfaith91** [comment]
@liggitt I think the round-tripping test failed due to different `selector` type for `autoscaling.Scale` and `extensions.Scale`. The `selector` field is a string for `autoscaling.Scale`, but it is `LabelSelector` type for `extensions.Scale`.

**Comment by liggitt** [comment]
it's not because it's a different type, it's because the fuzzed data is an invalid selector. making a custom fuzzer now that only allows valid selectors.

**Comment by liggitt** [comment]
Fixed fuzzing test, added a commit to enable scale subresources for apps/v1

File: test/integration/scale/scale_test.go [hunk 2]
```diff
@@ -68,9 +68,9 @@ func TestScaleSubresources(t *testing.T) {
 		makeGVR("apps", "v1beta2", "replicasets/scale"):  makeGVK("apps", "v1beta2", "Scale"),
 		makeGVR("apps", "v1beta2", "statefulsets/scale"): makeGVK("apps", "v1beta2", "Scale"),
 
-		// makeGVR("apps", "v1", "deployments/scale"):  makeGVK("autoscaling", "v1", "Scale"),
-		// makeGVR("apps", "v1", "replicasets/scale"):  makeGVK("autoscaling", "v1", "Scale"),
-		// makeGVR("apps", "v1", "statefulsets/scale"): makeGVK("autoscaling", "v1", "Scale"),
+		makeGVR("apps", "v1", "deployments/scale"):  makeGVK("autoscaling", "v1", "Scale"),
```
Comment (crimsonfaith91) [inline test/integration/scale/scale_test.go]: @liggitt Yes, this works! Thank you!

**Comment by liggitt** [comment]
cc @kubernetes/sig-apps-api-reviews @kubernetes/api-reviewers

File: pkg/apis/apps/v1beta1/conversion.go [hunk 3]
```diff
@@ -178,48 +180,35 @@ func Convert_apps_StatefulSetUpdateStrategy_To_v1beta1_StatefulSetUpdateStrategy
...
-		selector, err := metav1.LabelSelectorAsSelector(in.Selector)
-		if err != nil {
-			return fmt.Errorf("invalid label selector: %v", err)
-		}
-		out.TargetSelector = selector.String()
+	selector, err := metav1.ParseToLabelSelector(in.Selector)
+	if err != nil {
+		return fmt.Errorf("failed to parse selector: %v", err)
 	}
+	if len(selector.MatchExpressions) == 0 {
```
Comment (crimsonfaith91) [inline pkg/apis/apps/v1beta1/conversion.go]: Do we need to handle the case when `len(selector.MatchExpressions) != 0`?

File: pkg/apis/extensions/types.go [hunk 4]
```diff
@@ -42,44 +42,6 @@ const (
 	SysctlsPodSecurityPolicyAnnotationKey string = "security.alpha.kubernetes.io/sysctls"
 )
 
-// describes the attributes of a scale subresource
```
Comment (crimsonfaith91) [inline pkg/apis/extensions/types.go]: Just to verify - removing these will not break backward compatibility, right?

File: pkg/apis/apps/v1beta1/conversion.go (hunk 3 above)
Comment (liggitt) [reply pkg/apis/apps/v1beta1/conversion.go]: no, if `MatchExpressions` is non-empty, then `MatchLabels` is insufficient to express the full selector, and we should not set it.

This is the same logic as before:
```
if in.Selector.MatchExpressions == nil || len(in.Selector.MatchExpressions) == 0 {
	out.Selector = in.Selector.MatchLabels
}
```

File: pkg/apis/extensions/types.go (hunk 4 above)
Comment (liggitt) [reply pkg/apis/extensions/types.go]: Correct, as seen by there being zero changes in the openapi/swagger schemas (only apps/v1 additions), or external API types, or client-go methods.

File: pkg/registry/extensions/controller/storage/storage.go [hunk 5]
```diff
@@ -58,7 +60,7 @@ var _ = rest.Patcher(&ScaleREST{})
 
 // New creates a new Scale object
 func (r *ScaleREST) New() runtime.Object {
-	return &extensions.Scale{}
+	return &autoscaling.Scale{}
```
Comment (crimsonfaith91) [inline pkg/registry/extensions/controller/storage/storage.go]: @liggitt IMO we should not change any codes in `pkg/registry/extensions/controller/storage` directory as it is relevant to replication controller, which is v1.

Comment (liggitt) [reply pkg/registry/extensions/controller/storage/storage.go]: not an option, the internal `extensions.Scale` type is going away. This has no external API implications.

File: pkg/apis/apps/v1beta1/conversion.go (hunk 3 above)
Comment (crimsonfaith91) [reply pkg/apis/apps/v1beta1/conversion.go]: Good point, thanks!

File: pkg/registry/extensions/controller/storage/storage.go (hunk 5 above)
Comment (crimsonfaith91) [reply pkg/registry/extensions/controller/storage/storage.go]: sgtm :)

Comment (liggitt) [reply pkg/registry/extensions/controller/storage/storage.go]: Also, this is only used for `/apis/extensions/v1beta1/namespaces/ns/replicationcontrollers/rc/scale`, which is not how replication controllers should be scaled. The `/api/v1/namespaces/ns/replicationcontrollers/rc/scale` API is the v1 scale subresource for replication controllers

Comment (crimsonfaith91) [reply pkg/registry/extensions/controller/storage/storage.go]: noted! didn't know about that, thanks for pointing it out!

**Comment by liggitt** [comment]
/hold
would like ack on API review

**Review by kow3ns — APPROVED** [review]
LGTM 
IMO this fixes #49504
Can we add a release note that indicates the app/v1 group uses autoscaling/v1 Scale.

**Comment by liggitt** [comment]
> IMO this fixes #49504

this lays the groundwork for it, but until `kubectl scale` actually uses discovery and scale subresources consistently, that should stay open

> Can we add a release note that indicates the app/v1 group uses autoscaling/v1 Scale.

I can... since apps/v1 is new in this release, I figured we'd combine release notes for all the apps/v1 stuff into a single note, not per-PR

**Pull request merged**